rxtools/source/features/downgradeapp.c

```diff
@@ -11,169 +11,582 @@
     unsigned int titleid_low = 0x00040010;
     unsigned int titleid_high[] = { 0x00020000, 0x00021000, 0x00022000/*, 0x00026000, 0x00027000, 0x00028000*/}; //JPN, USA, EUR, CHN, KOR, TWN
-    char* regions[] = { "Japan", "USA", "Europe", /*"China", "Korea", "Taiwan", */"Unknown"};
     unsigned int mset_hash[] = { 0x57358F14, 0xA28EAD9F, 0x530C345B };   //JPN, USA, EUR
 	unsigned int supported_regions = sizeof(titleid_high)/sizeof(titleid_high[0]);
-    unsigned int region;
-    ConsoleInit();
-    ConsoleSetTitle("MSET Downgrader");
 	
+	ConsoleInit();
+    ConsoleSetTitle("MSET Downgrader");
     print("Opening MSET app...\n"); ConsoleShow();
-    for(region = 0; region < supported_regions && !FindApp(&filepath, titleid_low, titleid_high[region], "1"); region++);
-    if(region < supported_regions){
-        print("Region : %s\n\n", regions[region]); ConsoleShow();
-		//print(filepath); print("\n"); ConsoleShow(); FileCopy("mset.app", filepath);
-        FileOpen(&dg, filepath, 0);
-        unsigned int check_val;
-        FileRead(&dg, &check_val, 4, 0x130);
-        FileClose(&dg);
-        if(check_val == 0){
-            print("MSET is already in his exploitable\nversion.\nThere's no need to downgrade.\n"); ConsoleShow();
-        }else if(checkDgFile(dgpath, mset_hash[region])){
-                print("Opening Downgrade Pack...\n"); ConsoleShow();
-                FileOpen(&dg, dgpath, 0);
-                unsigned int dgsize = FileGetSize(&dg);
-                unsigned char* buf = 0x21000000;
-                FileRead(&dg, buf, dgsize, 0);
-                //Decrypting...
-                u8 iv[0x10] = {0};
-                u8 Key[0x10] = {0};
-                GetTitleKey(&Key, titleid_low, titleid_high[region]);
-            	aes_context aes_ctxt;
-            	aes_setkey_dec(&aes_ctxt, Key, 0x80);
-            	aes_crypt_cbc(&aes_ctxt, AES_DECRYPT, dgsize, iv, buf, buf);
-                FileWrite(&dg, buf, dgsize, 0);
-                FileClose(&dg);
-                //[/Decrypting]
-
-                if(*((unsigned int*)(buf + 0x100)) == 0x4843434E){
-                    print("Downgrading...\n"); ConsoleShow();
-                    FileCopy(filepath, dgpath);
-                    print("Done!\nRemoving Downgrade Pack...\n"); ConsoleShow();
-                    f_unlink(dgpath);
-                }else{
-                    print("Bad Downgrade Pack!\n"); ConsoleShow();
-                }
-            }else{
-                print("Cannot read Downgrade Pack!\n"); ConsoleShow();
-            }
-    }else{
-        print("Cannot read MSET app!\n"); ConsoleShow();
-    }
+	
+	if (CheckRegion("1") == 0)
+	{
+		if (region < supported_regions)
+		{
+			if (FindApp(titleid_low, titleid_high[region], "1")) // SysNAND only
+			{
+				FileOpen(&dg, cntpath, 0);
+				unsigned int check_val;
+				FileRead(&dg, &check_val, 4, 0x130);
+				FileClose(&dg);
+				
+				if (check_val != 0)
+				{
+					if (checkDgFile(dgpath, mset_hash[region]))
```

Comment:
About this, I prefer to wait until Roxas75 decides what's best. I'll update my code if he says the use of goto jumps is ok.